### ISSUE TITLE
Inconsistent Behavior in CheckoutDelegateWrapper Callbacks (onComplete, onCancel)

### ISSUE THREAD

**Issue by sumeyyekazanci** [Shopify/checkout-sheet-kit-swift#264]
### What area is the issue related to?

Checkout Sheet Kit

### What version of Checkout Sheet Kit are you using?

3.0.4-3.1.2

### Steps to Reproduce

- When the app state changes to inactive and then returns to the foreground, the onCancel callback does not get triggered.

- iOS 17.0 onCancel never works

- onComplete is not stable.

### Expected Behavior

The onComplete, onCancel, and other related callbacks should trigger reliably when the associated events occur.

### Actual Behavior

The Shopify Sheet Kit SDK does not work stably in certain scenarios, and specifically, callback functions such as onComplete and onCancel defined within the CheckoutDelegateWrapper behave inconsistently. This issue prevents users from receiving feedback during certain operations and causes problems in the application flow.

### Storefront domain

flyingtiger-tr.myshopify.com

**Comment by plump-pinniped** [comment]
Hey I was also running into this issue. For me, the issue happened when the view hosting the CheckoutSheet updated. I think it's because the CheckoutWebViewController only has a [weak ref to the delegate object](https://github.com/Shopify/checkout-sheet-kit-swift/blob/c858cbb26ad19e0248a32c9b938a30064730ce04/Sources/ShopifyCheckoutSheetKit/CheckoutWebViewController.swift#L28C1-L29C1). When the CheckoutSheet is updated, a new delegate is instantiated so the old one is deallocated. Setting a breakpoint [here](https://github.com/Shopify/checkout-sheet-kit-swift/blob/c858cbb26ad19e0248a32c9b938a30064730ce04/Sources/ShopifyCheckoutSheetKit/CheckoutWebViewController.swift#L154C1-L154C32), delegate was nil. I guess you'd have to update the delegate somehow in [updateUIViewController](https://github.com/Shopify/checkout-sheet-kit-swift/blob/c858cbb26ad19e0248a32c9b938a30064730ce04/Sources/ShopifyCheckoutSheetKit/CheckoutViewController.swift#L81C1-L82C1)

I did a workaround by making the host view Equatable based off of url so that the CheckoutSheet doesn't update. That fixed the issue for me

**Comment by kieran-osgood-shopify** [comment]
@randonnee @sumeyyekazanci - I have looked into this a bit today

Whilst I haven't been able to reproduce a case where the delegate goes to nil (if you are able to provide source code for a reproduction, I will look into it further), I have a PR [here](https://github.com/Shopify/checkout-sheet-kit-swift/pull/296) going along the lines you suggested @randonnee to ensure delegate is set when updateUIViewController is called.

If you fancy giving that a test and see if you still have any issues that'd be helpful before we look to merge.

On a related note, I have noticed that one of our [samples /CartView.swift](https://github.com/Shopify/checkout-sheet-kit-swift/blob/d319147ec5e4cb113f0b27a955aecd4b4d71fc19/Samples/SwiftUIExample/SwiftUIExample/CartView.swift#L63-L100) suggests using a sheet modifier to render the checkout sheet, and if the user were to swipe the sheet down the `onCancel` is never triggered, which may be a source of confusion.

```swift
    .sheet(isPresented: $isShowingCheckout) {
        if let url = CartManager.shared.cart?.checkoutUrl {
            let _ = print("[CartView] Creating CheckoutSheet with URL: \(url)")
            CheckoutSheet(checkout: url)
                .onCancel {
                    print("❌ CANCELLED")
                }
                .onComplete { _ in
                    print("✅ COMPLETE")
                }
                /// Configuration
                .title("SHEET")
        }
    }
```

**Comment by kieran-osgood-shopify** [comment]
I've just merged the above patch and released v3.2.0
I'll close this issue for now as theres not been any activity, but get back to us if you find you're still having issue
